lib/Driver/ToolChains.cpp

```diff
@@ -1493,6 +1534,128 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
   return II;
 }
 
+ToolChain::InvocationInfo
+toolchains::Windows::constructInvocation(const LinkJobAction &job,
+                                         const JobContext &context) const {
+  assert(context.Output.getPrimaryOutputType() == types::TY_Image &&
+         "Invalid linker output type.");
+
+  ArgStringList Arguments;
+
+  switch (job.getKind()) {
+  case LinkKind::None:
+    llvm_unreachable("invalid link kind");
+  case LinkKind::Executable:
+    // Default case, nothing extra needed.
+    break;
+  case LinkKind::DynamicLibrary:
+    Arguments.push_back("-shared");
+    break;
+  }
+
+  // Select the linker to use.
+  std::string Linker;
+  if (const Arg *A = context.Args.getLastArg(options::OPT_use_ld)) {
+    Linker = A->getValue();
+  }
+  if (!Linker.empty()) {
+    Arguments.push_back(context.Args.MakeArgString("-fuse-ld=" + Linker));
+  }
+
+  // Configure the toolchain.
+  // By default, use the system clang++ to link.
+  const char *Clang = "clang++";
+  if (const Arg *A = context.Args.getLastArg(options::OPT_tools_directory)) {
+    StringRef toolchainPath(A->getValue());
+
+    // If there is a clang in the toolchain folder, use that instead.
+    if (auto toolchainClang =
+            llvm::sys::findProgramByName("clang++", {toolchainPath})) {
+      Clang = context.Args.MakeArgString(toolchainClang.get());
+    }
+  }
+
+  std::string Target = getTriple().str();
+  if (!Target.empty()) {
+    Arguments.push_back("-target");
+    Arguments.push_back(context.Args.MakeArgString(Target));
+  }
+
+
+  SmallString<128> SharedRuntimeLibPath;
+  getRuntimeLibraryPath(SharedRuntimeLibPath, context.Args, *this);
+    
+  // Link the standard library.
+  Arguments.push_back("-L");
+  if (context.Args.hasFlag(options::OPT_static_stdlib,
+                            options::OPT_no_static_stdlib,
+                            false)) {
+    SmallString<128> StaticRuntimeLibPath;
+    getRuntimeStaticLibraryPath(StaticRuntimeLibPath, context.Args, *this);
+    
+    // Since Windows has separate libraries per architecture, link against the
+    // architecture specific version of the static library.
```

Comment:
Should `getRuntimeStaticLibraryPath` just do this? Are there places where we call it and need it to not do this?

Reply:
Windows is the only place we currently add the arch that I can see. I *think* on Darwin universal binaries are produced and copied to the lib/swift_static/<platform> folder; on Linux, I think the current assumption is that only one architecture will ever be present per platform, and so just using the copied libraries is fine (although probably not *technically* correct). 

On Windows, the files in that folder are (currently) just the static libraries for whichever arch was built last, so linking against that is definitely incorrect.

Reply:
So I think that's @zbowling's #13108. I think that pushes the arch up into `getRuntimeStaticLibraryPath` for everyone, so this would become redundant. But I'm not sure what state that PR is in, since we didn't get back to Zac very quickly and he hasn't touched it but then @alblue poked it again.

Reply:
That looks right. Once that's merged we can remove the extra code here and just use `getRuntimeStaticLibraryPathWithArch` instead.

Reply:
yeah, sorry. I've been busying in a bit of freeze with all the code I posted in all the PRs in December to take them and try and cut and stabilize a usable toolchain to ship in Fuchsia. Everything is getting to a good place with that toolchain so I should have more time to switch back and rebase all my PRs to get them ready to land upstream easier (and so I can get a new toolchain going with changes since the 4.1 branch).